src/main/java/io/cryostat/recordings/ActiveRecordings.java

```diff
@@ -119,18 +123,31 @@ public String patch(@RestPath long targetId, @RestPath long remoteId, String bod
                         .atMost(connectionFailedTimeout);
                 return null;
             case "save":
-                try {
-                    // FIXME this operation might take a long time to complete, depending on the
-                    // amount of JFR data in the target and the speed of the connection between the
-                    // target and Cryostat. We should not make the client wait until this operation
-                    // completes before sending a response - it should be async. Here we should just
-                    // return an Accepted response, and if a failure occurs that should be indicated
-                    // as a websocket notification.
-                    return recordingHelper.archiveRecording(activeRecording, null, null).name();
-                } catch (IOException ioe) {
-                    logger.warn(ioe);
-                    return null;
-                }
+                // FIXME this operation might take a long time to complete, depending on the
+                // amount of JFR data in the target and the speed of the connection between the
+                // target and Cryostat. We should not make the client wait until this operation
+                // completes before sending a response - it should be async. Here we should just
+                // return an Accepted response, and if a failure occurs that should be indicated
+                // as a websocket notification.
+
+                /*
+                 * Desired workflow:
+                 *   Client sends a PATCH request to Cryostat
+                 *   Cryostat receives the PATCH and checks that the specified active recording exists and that the target JVM is reachable (ex. try to open a connection and do something relatively lightweight like compute its JVM ID). If this check succeeds respond to the PATCH with 202, if it fails respond with a 404 (recording not found) or 502 (target not reachable) etc.
+                 *   In the background, Cryostat creates the S3 file upload request, opens a target connection, pipes the bytes, etc. - same as steps 2-5 above
+                 *   Cryostat emits a WebSocket notification, either indicating task successful completion or task failure.
+                 */
+                logger.info("Ceating request");
+                ArchiveRequest request =
+                        new ArchiveRequest(UUID.randomUUID().toString(), activeRecording);
+                logger.info(
+                        "Request created: ("
+                                + request.getId()
+                                + ", "
+                                + request.getRecording().name
+                                + ")");
+                bus.publish(ArchiveRequestGenerator.ARCHIVE_ADDRESS, request);
+                return request.getId();
```

Comment:
I think the `resp.addEndHandler(() -> bus.publish())` piece is important to help prevent thread timing issues. With this implementation it's possible that the generator request event gets put on the bus, then the generator services and completes that request and emits the websocket notification, and then the original worker thread continues execution and sends the UUID response back to the client after the completion notification has already been emitted. If this happens, then the client never knew what UUID to be listening for to know that their job was completed.

I guess this risk is almost always going to be present in some form with this general scheme. By using the end handler we expect to have written the UUID out on the original HTTP response to the client before starting the async generator job, but it's still possible (albeit very unlikely) that the actual network packets for the completion notification arrive at the client before the UUID response does.

Another possible scheme would be for the client to optionally supply a UUID of their own. If they do, then Cryostat should use that UUID and respond back to the client with the same UUID, and use that UUID for the async generator job too. This way it's guaranteed that the client knows what the job ID is before the job can possibly be completed. Otherwise, the execution flow would be the same.

Reply:
@Josh-Matsuoka any thoughts on this aspect?

Reply:
> resp.addEndHandler(() -> bus.publish())

Is there an example somewhere this piece is used? I couldn't find one in an initial search.

Reply:
There are no current examples in the codebase, just the reference from the other thread:

https://github.com/cryostatio/cryostat/pull/698#discussion_r1823248349

So I think it'll look like:

```diff
diff --git a/src/main/java/io/cryostat/recordings/ActiveRecordings.java b/src/main/java/io/cryostat/recordings/ActiveRecordings.java
index 3424eab6..aa6164a9 100644
--- a/src/main/java/io/cryostat/recordings/ActiveRecordings.java
+++ b/src/main/java/io/cryostat/recordings/ActiveRecordings.java
@@ -60,6 +60,7 @@ import org.jboss.resteasy.reactive.RestForm;
 import org.jboss.resteasy.reactive.RestPath;
 import org.jboss.resteasy.reactive.RestResponse;
 import org.jboss.resteasy.reactive.RestResponse.ResponseBuilder;
+import org.jboss.resteasy.reactive.server.spi.ServerHttpResponse;
 
 @Path("/api/v4/targets/{targetId}/recordings")
 public class ActiveRecordings {
@@ -104,7 +105,11 @@ public class ActiveRecordings {
     @Blocking
     @Path("/{remoteId}")
     @RolesAllowed("write")
-    public String patch(@RestPath long targetId, @RestPath long remoteId, String body)
+    public String patch(
+            ServerHttpResponse response,
+            @RestPath long targetId,
+            @RestPath long remoteId,
+            String body)
             throws Exception {
         Target target = Target.find("id", targetId).singleResult();
         Optional<ActiveRecording> recording =
@@ -146,7 +151,8 @@ public class ActiveRecordings {
                                 + ", "
                                 + request.getRecording().name
                                 + ")");
-                bus.publish(ArchiveRequestGenerator.ARCHIVE_ADDRESS, request);
+                response.addCloseHandler(
+                        () -> bus.publish(ArchiveRequestGenerator.ARCHIVE_ADDRESS, request));
                 return request.getId();
             default:
                 throw new BadRequestException(body);
```

Reply:
This approach makes sense to me to fix the race you highlighted, I'll give it a try. Thanks for the example